File: 01-arrays_&_hashing/00049-Group_Anagrams/00049-group_anagrams.py
# 49: Group Anagrams
# https://leetcode.com/problems/group-anagrams/

from collections import defaultdict

class Solution:
    def groupAnagrams(self, strs: list[str]) -> list[list[str]]:
        tmap = defaultdict(list)
        
        for s in strs:
            tmap[str(sorted(s))].append(s)
        
        return list(tmap.values())


if __name__ == "__main__":
    o = Solution()

    # INPUT
    strs = ["eat","tea","tan","ate","nat","bat"]

    # OUTPUT
    result = o.groupAnagrams(strs)
    print(result)

File: 01-arrays_&_hashing/00049-Group_Anagrams/00049-group_anagrams.go
// 49: Group Anagrams
// https://leetcode.com/problems/group-anagrams/

package main

import (
	"fmt"
	"sort"
	"strings"
)

// SOLUTION
func groupAnagrams(strs []string) [][]string {
	result := make([][]string, 0)
	tmap := make(map[string][]string, 0)
	
	for _, str := range strs {
		org := str
		ss := strings.Split(str, "")
		sort.Strings(ss)
		str = strings.Join(ss, "")
		tmap[str] = append(tmap[str], org)
	}

	for _, s := range tmap {
		result = append(result, s)
	}

	return result
}


func main() {
	// INPUT
	strs := []string{"eat","tea","tan","ate","nat","bat"}

	// OUTPUT
	result := groupAnagrams(strs)
	fmt.Println(result)
}
